<a href="https://colab.research.google.com/github/ShaunakSen/Deep-Learning/blob/master/1_Simple_Sentiment_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Simple Sentiment Analysis

[tutorial link](https://github.com/bentrevett/pytorch-sentiment-analysis/blob/master/1%20-%20Simple%20Sentiment%20Analysis.ipynb)


In this series we'll be building a machine learning model to detect sentiment (i.e. detect if a sentence is positive or negative) using PyTorch and TorchText. This will be done on movie reviews, using the IMDb dataset.

In this first notebook, we'll start very simple to understand the general concepts whilst not really caring about good results. Further notebooks will build on this knowledge and we'll actually get good results.

### Introduction

We'll be using a recurrent neural network (RNN) as they are commonly used in analysing sequences. An RNN takes in sequence of words, X={x1,...,xT}, one at a time, and produces a hidden state, h, for each word. We use the RNN recurrently by feeding in the current word xt as well as the hidden state from the previous word, ht−1, to produce the next hidden state, ht.

ht=RNN(xt,ht−1)
Once we have our final hidden state, hT, (from feeding in the last word in the sequence, xT) we feed it through a linear layer, f, (also known as a fully connected layer), to receive our predicted sentiment, ŷ =f(hT).

Below shows an example sentence, with the RNN predicting zero, which indicates a negative sentiment. The RNN is shown in orange and the linear layer shown in silver. Note that we use the same RNN for every word, i.e. it has the same parameters. The initial hidden state, h0, is a tensor initialized to all zeros.

![](https://nbviewer.jupyter.org/github/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment1.png)

Note: some layers and steps have been omitted from the diagram, but these will be explained later.

### Preparing Data

One of the main concepts of TorchText is the Field. These define how your data should be processed. In our sentiment classification task the data consists of both the raw string of the review and the sentiment, either "pos" or "neg".

The parameters of a Field specify how the data should be processed.

We use the TEXT field to define how the review should be processed, and the LABEL field to process the sentiment.

Our TEXT field has tokenize='spacy' as an argument. This defines that the "tokenization" (the act of splitting the string into discrete "tokens") should be done using the spaCy tokenizer. If no tokenize argument is passed, the default is simply splitting the string on spaces.

LABEL is defined by a LabelField, a special subset of the Field class specifically used for handling labels. We will explain the dtype argument later.

For more on Fields, go here.

We also set the random seeds for reproducibility.

In [0]:
import torch
from torchtext import data

SEED = 1234

torch.manual_seed(SEED)
torch.backends.cudnn.deterministic = True

TEXT = data.Field(tokenize='spacy')
LABEL = data.LabelField(dtype = torch.float)

Another handy feature of TorchText is that it has support for common datasets used in natural language processing (NLP).

The following code automatically downloads the IMDb dataset and splits it into the canonical train/test splits as torchtext.datasets objects. It process the data using the Fields we have previously defined. The IMDb dataset consists of 50,000 movie reviews, each marked as being a positive or negative review.

In [0]:
from torchtext import datasets

train_data, test_data = datasets.IMDB.splits(TEXT, LABEL)

In [56]:

print(f'Number of training examples: {len(train_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 25000
Number of testing examples: 25000


We can also check an example.



In [57]:
print(vars(train_data.examples[0]))


{'text': ['Now', 'let', 'me', 'tell', 'you', 'about', 'this', 'movie', ',', 'this', 'movie', 'is', 'MY', 'FAVORITE', 'MOVIE', '!', '!', '!', 'This', 'movie', 'has', 'excellent', 'combat', 'fighting', '.', 'This', 'movie', 'does', 'sound', 'like', 'a', 'silly', 'story', 'line', 'about', 'how', 'Jet', 'Li', 'plays', 'a', 'super', 'hero', ',', 'like', 'Spider', '-', 'Man', ',', 'or', 'etc', '.', 'But', 'once', 'you', "'ve", 'seen', 'this', 'movie', ',', 'you', 'would', 'probably', 'want', 'to', 'see', 'it', 'again', 'and', 'again', '.', 'I', 'rate', 'this', 'movie', '10/10', '.'], 'label': 'pos'}


The IMDb dataset only has train/test splits, so we need to create a validation set. We can do this with the .split() method.

By default this splits 70/30, however by passing a split_ratio argument, we can change the ratio of the split, i.e. a split_ratio of 0.8 would mean 80% of the examples make up the training set and 20% make up the validation set.

We also pass our random seed to the random_state argument, ensuring that we get the same train/validation split each time.

In [0]:
import random

train_data, valid_data = train_data.split(random_state = random.seed(SEED))

In [59]:
print(f'Number of training examples: {len(train_data)}')
print(f'Number of validation examples: {len(valid_data)}')
print(f'Number of testing examples: {len(test_data)}')

Number of training examples: 17500
Number of validation examples: 7500
Number of testing examples: 25000


In [60]:
print (vars(valid_data.examples[0]))
print (vars(test_data.examples[0]))

{'text': ['This', 'is', 'a', 'tongue', 'in', 'cheek', 'movie', 'from', 'the', 'very', 'outset', 'with', 'a', 'voice', '-', 'over', 'that', 'pokes', 'fun', 'at', 'everything', 'French', 'and', 'then', 'produces', 'a', 'rather', 'naif', 'but', 'very', 'brave', 'hero', 'in', 'Fanfan', 'La', 'Tulipe', '.', 'Portrayed', 'by', 'the', 'splendid', 'Gerard', 'Philippe', ',', 'the', 'dashing', 'young', 'man', 'believes', 'utterly', 'in', 'the', 'fate', 'curvaceous', 'Lollobrigida', 'foretells', '-', 'notably', 'that', 'he', 'will', 'marry', 'King', 'Louis', 'XV', "'s", 'daughter', '!', 'Problem', 'is', ',', 'La', 'Lollo', 'soon', 'find', 'outs', 'she', 'too', 'is', 'in', 'love', 'with', 'Fanfan', '...', '<br', '/><br', '/>Propelled', 'by', 'good', 'sword', 'fights', ',', 'cavalcades', ',', 'and', 'other', 'spirited', 'action', 'sequences', 'the', 'film', 'moves', 'at', 'a', 'brisk', 'pace', 'and', 'with', 'many', 'comic', 'moments', '.', 'The', 'direction', 'is', 'perhaps', 'the', 'weakest', 'as

Next, we have to build a vocabulary. This is a effectively a look up table where every unique word in your data set has a corresponding index (an integer).

We do this as our machine learning model cannot operate on strings, only numbers. Each index is used to construct a one-hot vector for each word. A one-hot vector is a vector where all of the elements are 0, except one, which is 1, and dimensionality is the total number of unique words in your vocabulary, commonly denoted by V.

![](https://nbviewer.jupyter.org/github/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment5.png)

The number of unique words in our training set is over 100,000, which means that our one-hot vectors will have over 100,000 dimensions! This will make training slow and possibly won't fit onto your GPU (if you're using one).

There are two ways effectively cut down our vocabulary, we can either only take the top n most common words or ignore words that appear less than m times. We'll do the former, only keeping the top 25,000 words.

What do we do with words that appear in examples but we have cut from the vocabulary? We replace them with a special unknown or <unk> token. For example, if the sentence was "This film is great and I love it" but the word "love" was not in the vocabulary, it would become "This film is great and I <unk> it".

The following builds the vocabulary, only keeping the most common max_size tokens.



In [0]:
MAX_VOCAB_SIZE = 25_000

TEXT.build_vocab(train_data, max_size=MAX_VOCAB_SIZE)
LABEL.build_vocab(train_data)

Why do we only build the vocabulary on the training set? When testing any machine learning system you do not want to look at the test set in any way. We do not include the validation set as we want it to reflect the test set as much as possible.



In [62]:
print(f"Unique tokens in TEXT vocabulary: {len(TEXT.vocab)}")
print(f"Unique tokens in LABEL vocabulary: {len(LABEL.vocab)}")

Unique tokens in TEXT vocabulary: 25002
Unique tokens in LABEL vocabulary: 2


Why is the vocab size 25002 and not 25000? One of the addition tokens is the <unk> token and the other is a <pad> token.

When we feed sentences into our model, we feed a batch of them at a time, i.e. more than one at a time, and all sentences in the batch need to be the same size. Thus, to ensure each sentence in the batch is the same size, any shorter than the longest within the batch are padded.
  
![](https://nbviewer.jupyter.org/github/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment6.png)
  
We can also view the most common words in the vocabulary and their frequencies.



In [63]:
print(TEXT.vocab.freqs.most_common(20))

[('the', 201510), (',', 191197), ('.', 164359), ('and', 108899), ('a', 108393), ('of', 99722), ('to', 92747), ('is', 75886), ('in', 60851), ('I', 53729), ('it', 53382), ('that', 48976), ('"', 44204), ("'s", 43134), ('this', 42087), ('-', 36850), ('/><br', 35315), ('was', 35040), ('as', 30372), ('with', 29774)]


We can also see the vocabulary directly using either the stoi (string to int) or itos (int to string) method.



In [64]:
print(TEXT.vocab.itos)
print(TEXT.vocab.stoi)

['<unk>', '<pad>', 'the', ',', '.', 'and', 'a', 'of', 'to', 'is', 'in', 'I', 'it', 'that', '"', "'s", 'this', '-', '/><br', 'was', 'as', 'with', 'movie', 'for', 'film', 'The', 'but', '(', ')', "n't", 'on', 'you', 'are', 'not', 'his', 'have', 'be', 'he', 'one', 'at', 'by', '!', 'all', 'an', 'who', 'from', 'like', 'they', 'so', "'", 'her', 'has', 'about', 'or', 'out', 'It', 'just', 'do', '?', 'some', 'good', 'more', 'would', 'very', 'up', 'This', 'what', 'there', 'time', 'can', 'which', 'when', 'had', 'only', 'if', 'really', 'story', 'she', 'were', 'even', 'see', 'their', 'no', 'my', 'did', 'does', 'me', 'than', '...', 'much', ':', 'been', 'could', 'into', 'well', 'get', 'will', 'other', 'bad', 'him', 'people', 'because', 'we', 'great', 'made', 'most', 'first', 'also', '<', 'make', 'br', 'way', 'how', 'them', 'its', 'movies', 'any', 'too', '/>The', 'think', 'character', 'characters', 'then', 'seen', 'films', 'But', ';', 'plot', 'two', 'watch', 'being', 'many', 'acting', 'life', 'never', 

We can also check the labels, ensuring 0 is for negative and 1 is for positive.



In [65]:
print (LABEL.vocab.stoi)

defaultdict(<function _default_unk_index at 0x7f8aa196fc80>, {'neg': 0, 'pos': 1})


The final step of preparing the data is creating the iterators. We iterate over these in the training/evaluation loop, and they return a batch of examples (indexed and converted into tensors) at each iteration.

We'll use a BucketIterator which is a special type of iterator that will return a batch of examples where each example is of a similar length, minimizing the amount of padding per example.

We also want to place the tensors returned by the iterator on the GPU (if you're using one). PyTorch handles this using torch.device, we then pass this device to the iterator.



In [0]:
BATCH_SIZE = 64

# Create Iterator objects for multiple splits of a dataset

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

train_iterator, valid_iterator, test_iterator = data.BucketIterator.splits(
    datasets=(train_data, valid_data, test_data), 
    batch_size = BATCH_SIZE,
    device = device)

### Build the model

The next stage is building the model that we'll eventually train and evaluate.

There is a small amount of boilerplate code when creating models in PyTorch, note how our RNN class is a sub-class of nn.Module and the use of super.

Within the __init__ we define the layers of the module. Our three layers are an embedding layer, our RNN, and a linear layer. All layers have their parameters initialized to random values, unless explicitly specified.

The embedding layer is used to transform our sparse one-hot vector (sparse as most of the elements are 0) into a dense embedding vector (dense as the dimensionality is a lot smaller and all the elements are real numbers). This embedding layer is simply a single fully connected layer. As well as reducing the dimensionality of the input to the RNN, there is the theory that words which have similar impact on the sentiment of the review are mapped close together in this dense vector space. For more information about word embeddings, see here.

The RNN layer is our RNN which takes in our dense vector and the previous hidden state ht−1, which it uses to calculate the next hidden state, ht.

![](https://nbviewer.jupyter.org/github/bentrevett/pytorch-sentiment-analysis/blob/master/assets/sentiment7.png)

Finally, the linear layer takes the final hidden state and feeds it through a fully connected layer, f(hT), transforming it to the correct output dimension.

The forward method is called when we feed examples into our model.

Each batch, text, is a tensor of size [sentence length, batch size]. That is a batch of sentences, each having each word converted into a one-hot vector.

You may notice that this tensor should have another dimension due to the one-hot vectors, however PyTorch conveniently stores a one-hot vector as it's index value, i.e. the tensor representing a sentence is just a tensor of the indexes for each token in that sentence. The act of converting a list of tokens into a list of indexes is commonly called numericalizing.

The input batch is then passed through the embedding layer to get embedded, which gives us a dense vector representation of our sentences. embedded is a tensor of size [sentence length, batch size, embedding dim].

embedded is then fed into the RNN. In some frameworks you must feed the initial hidden state, h0, into the RNN, however in PyTorch, if no initial hidden state is passed as an argument it defaults to a tensor of all zeros.

The RNN returns 2 tensors, output of size [sentence length, batch size, hidden dim] and hidden of size [1, batch size, hidden dim]. output is the concatenation of the hidden state from every time step, whereas hidden is simply the final hidden state. We verify this using the assert statement. Note the squeeze method, which is used to remove a dimension of size 1.

Finally, we feed the last hidden state, hidden, through the linear layer, fc, to produce a prediction.

In [67]:
print(vars(train_data.examples[0])['text'])

['Film', 'version', 'of', 'Sandra', 'Bernhard', "'s", 'one', '-', 'woman', 'off', '-', 'Broadway', 'show', 'is', 'gaspingly', 'pretentious', '.', 'Sandra', 'spoofs', 'lounge', 'acts', 'and', 'superstars', ',', 'but', 'her', 'sense', 'of', 'irony', 'is', 'only', 'fitfully', 'interesting', ',', 'and', 'fitfully', 'funny', '.', 'Her', 'fans', 'will', 'say', 'she', "'s", 'scathingly', 'honest', ',', 'and', 'that', 'may', 'be', 'true', '.', 'But', 'she', "'s", 'also', 'shrill', ',', 'with', 'an', 'unapologetic', ',', 'in', '-', 'your', '-', 'face', 'bravado', 'that', 'is', "n't", 'well', '-', 'suited', 'to', 'a', 'film', 'in', 'this', 'genre', '.', 'She', 'does', "n't", 'want', 'to', 'make', 'nice', '--', 'and', 'she', "'s", 'certainly', 'not', 'out', 'to', 'make', 'friends', '--', 'and', 'that', "'s", 'always', 'going', 'to', 'rub', 'a', 'lot', 'of', 'people', 'the', 'wrong', 'way', '.', 'But', 'even', 'if', 'you', 'meet', 'her', 'halfway', ',', 'her', 'material', 'here', 'is', 'seriously'

In [102]:
# counting number of batches
count = 0
for x in train_iterator:
  count+=1
  # print(x.text.shape)
  #print(x.label)
  
print (count)

274


In [69]:
# there is a batch of size=28
274*64 - (64-28)

17500

In [73]:
my_list = [1,2,3]

my_list[-1]

3

In [0]:
import torch.nn as nn

class RNN(nn.Module):
  def __init__(self, input_dim, embedding_dim, hidden_dim, output_dim):
    super().__init__()
    self.embedding = nn.Embedding(input_dim, embedding_dim)
    
    # rnn should take embedding dim and prev hidden state (h_t-1)  which it uses to calculate the next hidden state, h_t
    self.rnn = nn.RNN(embedding_dim, hidden_dim)
    self.fc = nn.Linear(hidden_dim, output_dim)
    
  def forward(self, text):
    #text = [sent len, batch size]
    embedded = self.embedding(text)
    
    #embedded = [sent len, batch size, emb dim]
    
    # The RNN returns 2 tensors, output of size [sentence length, batch size, hidden dim] and hidden of size [1, batch size, hidden dim]
    
    output, hidden = self.rnn(embedded)
    
    #output = [sent len, batch size, hid dim]
    #hidden = [1, batch size, hid dim]
    
    # hidden.squeeze(0) -> [batch_size, hid_dim]
    
    # output is the concatenation of the hidden state from every time step, whereas hidden is simply the final hidden state.
    
    assert torch.equal(output[-1,:,:], hidden.squeeze(dim=0))
        
    return self.fc(hidden.squeeze(dim=0))


We now create an instance of our RNN class.

The input dimension is the dimension of the one-hot vectors, which is equal to the vocabulary size.

The embedding dimension is the size of the dense word vectors. This is usually around 50-250 dimensions, but depends on the size of the vocabulary.

The hidden dimension is the size of the hidden states. This is usually around 100-500 dimensions, but also depends on factors such as on the vocabulary size, the size of the dense vectors and the complexity of the task.

The output dimension is usually the number of classes, however in the case of only 2 classes the output value is between 0 and 1 and thus can be 1-dimensional, i.e. a single scalar real number.

In [0]:
INPUT_DIM = len(TEXT.vocab)
EMBEDDING_DIM = 100
HIDDEN_DIM = 256
OUTPUT_DIM = 1

model = RNN(INPUT_DIM, EMBEDDING_DIM, HIDDEN_DIM, OUTPUT_DIM)

Let's also create a function that will tell us how many trainable parameters our model has so we can compare the number of parameters across different models.



In [80]:
def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

print(f'The model has {count_parameters(model):,} trainable parameters')

The model has 2,592,105 trainable parameters


### Train the Model


Now we'll set up the training and then train the model.

First, we'll create an optimizer. This is the algorithm we use to update the parameters of the module. Here, we'll use stochastic gradient descent (SGD). The first argument is the parameters will be updated by the optimizer, the second is the learning rate, i.e. how much we'll change the parameters by when we do a parameter update.

In [0]:
import torch.optim as optim

optimizer = optim.SGD(params=model.parameters(), lr=1e-3)

Next, we'll define our loss function. In PyTorch this is commonly called a criterion.

The loss function here is binary cross entropy with logits.

Our model currently outputs an unbound real number. As our labels are either 0 or 1, we want to restrict the predictions to a number between 0 and 1. We do this using the sigmoid or logit functions.

We then use this this bound scalar to calculate the loss using binary cross entropy.

The BCEWithLogitsLoss criterion carries out both the sigmoid and the binary cross entropy steps.

In [0]:
criterion = nn.BCEWithLogitsLoss()

Using .to, we can place the model and the criterion on the GPU (if we have one).



In [0]:
model = model.to(device)
criterion = criterion.to(device)

Our criterion function calculates the loss, however we have to write our function to calculate the accuracy.

This function first feeds the predictions through a sigmoid layer, squashing the values between 0 and 1, we then round them to the nearest integer. This rounds any value greater than 0.5 to 1 (a positive sentiment) and the rest to 0 (a negative sentiment).

We then calculate how many rounded predictions equal the actual labels and average it across the batch.

In [99]:
def binary_accuracy(preds, y):
  """
    Returns accuracy per batch, i.e. if you get 8/10 right, this returns 0.8, NOT 8
  """
  #round predictions to the closest integer
  
  # print (torch.sigmoid(preds))
  
  rounded_preds = torch.round(torch.sigmoid(preds))
  
  # rounded_preds == y returns [1,0,0,1... ] 1 if equal else 0
  
  correct = (rounded_preds == y).float() #convert into float for division 
  
  accuracy = correct.sum()/len(correct)
  
  return accuracy


print(binary_accuracy(torch.tensor([2,3,0.3,-1]), torch.tensor([1.,0,0,0])))

tensor(0.5000)


The train function iterates over all examples, one batch at a time.

model.train() is used to put the model in "training mode", which turns on dropout and batch normalization. Although we aren't using them in this model, it's good practice to include it.

For each batch, we first zero the gradients. Each parameter in a model has a grad attribute which stores the gradient calculated by the criterion. PyTorch does not automatically remove (or "zero") the gradients calculated from the last gradient calculation, so they must be manually zeroed.

We then feed the batch of sentences, batch.text, into the model. Note, you do not need to do model.forward(batch.text), simply calling the model works. The squeeze is needed as the predictions are initially size [batch size, 1], and we need to remove the dimension of size 1 as PyTorch expects the predictions input to our criterion function to be of size [batch size].

The loss and accuracy are then calculated using our predictions and the labels, batch.label, with the loss being averaged over all examples in the batch.

We calculate the gradient of each parameter with loss.backward(), and then update the parameters using the gradients and optimizer algorithm with optimizer.step().

The loss and accuracy is accumulated across the epoch, the .item() method is used to extract a scalar from a tensor which only contains a single value.

Finally, we return the loss and accuracy, averaged across the epoch. The len of an iterator is the number of batches in the iterator.

You may recall when initializing the LABEL field, we set dtype=torch.float. This is because TorchText sets tensors to be LongTensors by default, however our criterion expects both inputs to be FloatTensors. Setting the dtype to be torch.float, did this for us. The alternative method of doing this would be to do the conversion inside the train function by passing batch.label.float() instad of batch.label to the criterion.

In [0]:
def train(model, iterator, optimizer, criterion):
  
  epoch_loss = 0
  epoch_acc = 0

  model.train()

  for batch in iterator:

      optimizer.zero_grad()

      predictions = model(batch.text).squeeze(1)

      loss = criterion(predictions, batch.label)

      acc = binary_accuracy(predictions, batch.label)

      loss.backward()

      optimizer.step()

      epoch_loss += loss.item()
      epoch_acc += acc.item()

  return epoch_loss / len(iterator), epoch_acc / len(iterator)

evaluate is similar to train, with a few modifications as you don't want to update the parameters when evaluating.

model.eval() puts the model in "evaluation mode", this turns off dropout and batch normalization. Again, we are not using them in this model, but it is good practice to include them.

No gradients are calculated on PyTorch operations inside the with no_grad() block. This causes less memory to be used and speeds up computation.

The rest of the function is the same as train, with the removal of optimizer.zero_grad(), loss.backward() and optimizer.step(), as we do not update the model's parameters when evaluating.

In [0]:
def evaluate(model, iterator, criterion):
    
    epoch_loss = 0
    epoch_acc = 0
    
    model.eval()
    
    with torch.no_grad():
    
        for batch in iterator:

            predictions = model(batch.text).squeeze(1)
            
            loss = criterion(predictions, batch.label)
            
            acc = binary_accuracy(predictions, batch.label)

            epoch_loss += loss.item()
            epoch_acc += acc.item()
        
    return epoch_loss / len(iterator), epoch_acc / len(iterator)

We'll also create a function to tell us how long an epoch takes to compare training times between models.



In [0]:
import time

def epoch_time(start_time, end_time):
    elapsed_time = end_time - start_time
    elapsed_mins = int(elapsed_time / 60)
    elapsed_secs = int(elapsed_time - (elapsed_mins * 60))
    return elapsed_mins, elapsed_secs

We then train the model through multiple epochs, an epoch being a complete pass through all examples in the training and validation sets.

At each epoch, if the validation loss is the best we have seen so far, we'll save the parameters of the model and then after training has finished we'll use that model on the test set.

In [106]:
N_EPOCHS = 5

best_valid_loss = float('inf')

for epoch in range(N_EPOCHS):
  start_time = time.time()
  
  train_loss, train_acc = train(model, train_iterator, optimizer, criterion)
  valid_loss, valid_acc = evaluate(model, valid_iterator, criterion)

  end_time = time.time()

  epoch_mins, epoch_secs = epoch_time(start_time, end_time)
  
  if valid_loss < best_valid_loss:
    best_valid_loss = valid_loss
    torch.save(model.state_dict(), 'tut1-model.pt')
  print(f'Epoch: {epoch+1:02} | Epoch Time: {epoch_mins}m {epoch_secs}s')
  print(f'\tTrain Loss: {train_loss:.3f} | Train Acc: {train_acc*100:.2f}%')
  print(f'\t Val. Loss: {valid_loss:.3f} |  Val. Acc: {valid_acc*100:.2f}%')
        
    

Epoch: 01 | Epoch Time: 0m 23s
	Train Loss: 0.694 | Train Acc: 50.03%
	 Val. Loss: 0.697 |  Val. Acc: 48.99%
Epoch: 02 | Epoch Time: 0m 23s
	Train Loss: 0.693 | Train Acc: 50.11%
	 Val. Loss: 0.697 |  Val. Acc: 50.51%
Epoch: 03 | Epoch Time: 0m 24s
	Train Loss: 0.693 | Train Acc: 49.68%
	 Val. Loss: 0.697 |  Val. Acc: 49.05%
Epoch: 04 | Epoch Time: 0m 23s
	Train Loss: 0.693 | Train Acc: 49.85%
	 Val. Loss: 0.697 |  Val. Acc: 49.11%
Epoch: 05 | Epoch Time: 0m 23s
	Train Loss: 0.693 | Train Acc: 50.22%
	 Val. Loss: 0.697 |  Val. Acc: 49.05%


You may have noticed the loss is not really decreasing and the accuracy is poor. This is due to several issues with the model which we'll improve in the next notebook.

Finally, the metric we actually care about, the test loss and accuracy, which we get from our parameters that gave us the best validation loss.



In [107]:
model.load_state_dict(torch.load('tut1-model.pt'))

test_loss, test_acc = evaluate(model, test_iterator, criterion)

print(f'Test Loss: {test_loss:.3f} | Test Acc: {test_acc*100:.2f}%')

Test Loss: 0.712 | Test Acc: 46.57%
